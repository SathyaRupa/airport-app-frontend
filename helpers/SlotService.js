import axios from 'axios';
import url from './Url';

export default {
  fetchAll: async (page, isAvailable) => {
    try {
      const response = await axios.get(
        `${url}/slots?page=${page}&is_available=${isAvailable}`,
      );
      return response.data;
    } catch (error) {
      console.log('Error fetching data: ', error);
    }
  },
};
