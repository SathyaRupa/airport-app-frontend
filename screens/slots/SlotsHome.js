import React from 'react';
import moment from 'moment';
import {useEffect, useState} from 'react';
import {FlatList, ActivityIndicator, View, StyleSheet} from 'react-native';
import {Avatar, Chip} from 'react-native-paper';
import {useIsFocused} from '@react-navigation/native';
import SlotService from '../../helpers/SlotService';
import SlotCard from './SlotCard';

function SlotsHome({navigation}) {
  const [allDataLoaded, setAllDataLoaded] = useState(false);
  const [fetchedPages, setFetchedPages] = useState([]);
  const [slots, setSlots] = useState([]);
  const [loading, setLoading] = useState(true);
  const [page, setPage] = useState(0);
  const [available, setAvailable] = useState(true);
  const isFocused = useIsFocused();

  useEffect(() => {
    async function fetchData() {
      if (!allDataLoaded && isFocused && !fetchedPages.includes(page)) {
        const response = await SlotService.fetchAll(page, available);
        setFetchedPages(prevPages => [...prevPages, page]);
        if (response.length === 0) {
          setAllDataLoaded(true);
        } else {
          setSlots(prevGates => [...prevGates, ...response]);
        }
      }
      setLoading(false);
    }
    fetchData();
  }, [page, allDataLoaded, isFocused, available]);

  const loadMoreSlots = () => {
    if (!allDataLoaded) {
      setPage(prevPage => prevPage + 1);
    }
  };

  const renderFooter = () => {
    return loading ? <ActivityIndicator size="large" color="#4D869C" /> : null;
  };

  const formatData = slotData => {
    const DATE_TIME_FORMAT = 'HH:mm DD-MM-YYYY';
    const formatedData = {
      startTime: moment(slotData.start_time).format(DATE_TIME_FORMAT),
      endTime: moment(slotData.end_time).format(DATE_TIME_FORMAT),
      available: slotData.status,
    };
    return formatedData;
  };

  const icon = (
    <Avatar.Image
      style={{backgroundColor: 'transparent'}}
      size={50}
      source={require('../../assets/icons/slots.png')}
    />
  );

  return (
    <View>
      <View style={styles.chipContainer}>
        <Chip
          style={[styles.chip, !available ? styles.checked : styles.unchecked]}
          icon="check"
          textStyle={styles.text}
          onPress={() => {
            setSlots([]);
            setFetchedPages([]);
            setPage(0);
            setAvailable(!available);
            setAllDataLoaded(false);
          }}>
          Display all
        </Chip>
      </View>

      <FlatList
        data={slots}
        renderItem={itemData => (
          <SlotCard
            id={itemData.item.id}
            value={formatData(itemData.item)}
            icon={icon}
            testId={`slot-card-${itemData.index}`}
          />
        )}
        onEndReached={loadMoreSlots}
        onEndReachedThreshold={0.5}
        ListFooterComponent={renderFooter}
      />
    </View>
  );
}

export default SlotsHome;

const styles = StyleSheet.create({
  chipContainer: {
    padding: 12,
    flexDirection: 'row-reverse',
  },
  chip: {
    width: 140,
    backgroundColor: '#7AB2B2',
  },
  checked: {
    borderWidth: 2,
    borderColor: 'black',
  },
  text: {
    color: 'white',
  },
});
