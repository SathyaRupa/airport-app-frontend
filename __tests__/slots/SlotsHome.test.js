import {fireEvent, render, waitFor} from '@testing-library/react-native';
import React from 'react';
import SlotsHome from '../../screens/slots/SlotsHome';
import SlotService from '../../helpers/SlotService';
import {useIsFocused} from '@react-navigation/native';

jest.mock('react-native-linear-gradient', () => 'LinearGradient');
jest.mock('@react-navigation/native', () => ({
  useIsFocused: jest.fn(),
}));
jest.mock('../../helpers/SlotService', () => ({
  fetchAll: jest.fn(),
}));

const slot1 = {
  id: '1',
  start_time: '2024-05-11T07:19:27.265575Z',
  end_time: '2024-05-31T13:08:03.019702Z',
  status: 'Booked',
  aircraft_id: '2',
  gate_id: '1',
};
const slot2 = {
  id: '2',
  start_time: '2024-05-26T01:13:28.813964Z',
  end_time: '2024-05-31T13:25:29.923002Z',
  status: 'Reserved',
  aircraft_id: '4',
  gate_id: '5',
};
const slot3 = {
  id: '3',
  start_time: '2024-05-26T01:13:28.813964Z',
  end_time: '2024-05-25T04:25:29.923002Z',
  status: 'Available',
  aircraft_id: '',
  gate_id: '',
};

describe('SlotsHome page', () => {
  let mockNavigaton;

  beforeEach(() => {
    mockNavigaton = {
      push: jest.fn(),
    };
    jest.clearAllMocks();
    jest.useFakeTimers();
  });

  it('snapshot test', () => {
    const slotsHome = render(<SlotsHome navigation={mockNavigaton} />);

    expect(slotsHome).toMatchSnapshot();
  });

  it('should render slot-card when data is recieved', () => {
    SlotService.fetchAll.mockResolvedValueOnce([slot1, slot2, slot3]);
    useIsFocused.mockReturnValue(true);

    const {queryByTestId} = render(<SlotsHome navigation={mockNavigaton} />);

    waitFor(() => {
      expect(queryByTestId('slot-card-0')).toBeTruthy();
      expect(queryByTestId('slot-card-1')).toBeTruthy();
      expect(queryByTestId('slot-card-2')).toBeTruthy();
      expect(SlotService.fetchAll).toHaveBeenCalledWith(0, true);
    });
  });

  it('should render call api again when checkbox is clicked', () => {
    SlotService.fetchAll.mockReturnValue([slot1, slot2, slot3]);
    useIsFocused.mockReturnValue(true);

    const {queryByTestId} = render(<SlotsHome navigation={mockNavigaton} />);
    const chip = queryByTestId('display-all');

    expect(chip).toBeTruthy();

    fireEvent.press(chip);
    waitFor(() => {
      expect(queryByTestId('slot-card-0')).toBeTruthy();
      expect(queryByTestId('slot-card-1')).toBeTruthy();
      expect(queryByTestId('slot-card-2')).toBeTruthy();
      expect(SlotService.fetchAll).toHaveBeenCalledWith(0, true);
      expect(SlotService.fetchAll).toHaveBeenCalledWith(0, false);
    });
  });
});
