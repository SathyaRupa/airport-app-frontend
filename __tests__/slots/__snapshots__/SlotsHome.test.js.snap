// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SlotsHome page snapshot test 1`] = `
<View>
  <View
    style={
      {
        "flexDirection": "row-reverse",
        "padding": 12,
      }
    }
  >
    <View
      collapsable={false}
      style={
        {
          "backgroundColor": "#7AB2B2",
          "borderRadius": 8,
          "shadowColor": "#000",
          "shadowOffset": {
            "height": 0,
            "width": 0,
          },
          "shadowOpacity": 0,
          "shadowRadius": 0,
          "width": 140,
        }
      }
      testID="display-all-container-outer-layer"
    >
      <View
        collapsable={false}
        style={
          {
            "backgroundColor": "#7AB2B2",
            "borderColor": "rgba(121, 116, 126, 1)",
            "borderRadius": 8,
            "borderStyle": "solid",
            "borderWidth": 0,
            "flex": undefined,
            "flexDirection": "column",
            "shadowColor": "#000",
            "shadowOffset": {
              "height": 0,
              "width": 0,
            },
            "shadowOpacity": 0,
            "shadowRadius": 0,
          }
        }
        testID="display-all-container"
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": false,
              "expanded": undefined,
              "selected": false,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            [
              {
                "overflow": "hidden",
              },
              [
                {
                  "borderRadius": 8,
                },
                {
                  "width": "100%",
                },
              ],
            ]
          }
          testID="display-all"
        >
          <View
            style={
              [
                {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "paddingLeft": 4,
                  "position": "relative",
                },
                {
                  "paddingLeft": 0,
                },
                {
                  "paddingRight": 0,
                },
              ]
            }
          >
            <View
              style={
                [
                  {
                    "alignSelf": "center",
                    "padding": 4,
                  },
                  {
                    "paddingLeft": 8,
                    "paddingRight": 0,
                  },
                  null,
                ]
              }
            >
              <Text
                accessibilityElementsHidden={true}
                importantForAccessibility="no-hide-descendants"
                pointerEvents="none"
                selectable={false}
                style={
                  [
                    {
                      "backgroundColor": "transparent",
                    },
                    {
                      "color": "rgba(103, 80, 164, 1)",
                      "fontSize": 18,
                    },
                  ]
                }
              >
                â–¡
              </Text>
            </View>
            <Text
              numberOfLines={1}
              selectable={false}
              style={
                [
                  {
                    "textAlign": "left",
                  },
                  {
                    "color": "rgba(28, 27, 31, 1)",
                    "writingDirection": "ltr",
                  },
                  [
                    {
                      "fontFamily": "System",
                      "fontSize": 14,
                      "fontWeight": "500",
                      "letterSpacing": 0.1,
                      "lineHeight": 20,
                    },
                    [
                      {
                        "marginVertical": 6,
                        "textAlignVertical": "center",
                      },
                      {
                        "color": "rgba(29, 25, 43, 1)",
                        "fontFamily": "System",
                        "fontSize": 14,
                        "fontWeight": "500",
                        "letterSpacing": 0.1,
                        "lineHeight": 20,
                      },
                      {
                        "marginLeft": 8,
                        "marginRight": 16,
                      },
                      {
                        "color": "white",
                      },
                    ],
                  ],
                ]
              }
            >
              Display all
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
  <RCTScrollView
    ListFooterComponent={[Function]}
    data={[]}
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onEndReached={[Function]}
    onEndReachedThreshold={0.5}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={0.0001}
    stickyHeaderIndices={[]}
    viewabilityConfigCallbackPairs={[]}
  >
    <View>
      <View
        onLayout={[Function]}
      />
    </View>
  </RCTScrollView>
</View>
`;
